token-swap/program/src/curve/stable.rs

```diff
@@ -82,53 +107,50 @@ fn compute_d(amp: u128, amount_a: u128, amount_b: u128) -> Option<u128> {
                 .checked_div(amount_b_times_coins)?;
             d_previous = d;
             //d = (leverage * sum_x + d_p * n_coins) * d / ((leverage - 1) * d + (n_coins + 1) * d_p);
-            d = calculate_step(d, leverage, sum_x, d_product, n_coins)?;
+            d = calculate_step(&d, leverage, sum_x, &d_product)?;
             // Equality with the precision of 1
-            if d > d_product {
-                if d.checked_sub(d_previous)? <= 1 {
-                    break;
-                }
-            } else if d_previous.checked_sub(d)? <= 1 {
+            if almost_equal(&d, &d_previous) {
                 break;
             }
         }
-        Some(d)
+        Some(d.as_u128())
     }
 }
 
 /// Compute swap amount `y` in proportion to `x`
 /// Solve for y:
 /// y**2 + y * (sum' - (A*n**n - 1) * D / (A * n**n)) = D ** (n + 1) / (n ** (2 * n) * prod' * A)
 /// y**2 + b*y = c
-fn compute_new_destination_amount(amp: u128, new_source_amount: u128, d_val: u128) -> Option<u128> {
-    // XXX: Curve uses u256
-    let n_coins: u128 = 2;
-    let leverage = amp.checked_mul(n_coins)?; // A * n
+fn compute_new_destination_amount(
+    leverage: u64,
+    new_source_amount: u128,
+    d_val: u128,
+) -> Option<u128> {
+    // Upscale to U256
+    let leverage: U256 = leverage.into();
+    let new_source_amount: U256 = new_source_amount.into();
+    let d_val: U256 = d_val.into();
 
     // sum' = prod' = x
     // c =  D ** (n + 1) / (n ** (2 * n) * prod' * A)
-    let c = checked_pow(d_val, n_coins.checked_add(1)? as usize)?.checked_div(
-        new_source_amount.checked_mul(checked_pow(n_coins, 2)?.checked_mul(leverage)?)?,
-    )?;
+    let c = checked_u8_power(&d_val, N_COINS + 1)?
```

Comment:
This is pedantic, but it helps with security reviews later -- can you change `N_COINS.checked_add(1)?`?

Reply:
Well, this is a constant now and its value is 2, do we really need checked_add there? No problem adding it, just curious :)

Reply:
I know it's totally pedantic, but it will absolutely help with security reviews